### DISCUSSION TITLE
STM32F411, how to drive PA0 for wake-up?

### DISCUSSION THREAD

**Discussion by davefes** [micropython/micropython#15811] [STM32 / Pyboard]
PA0 can not be low during boot and a high brings it out of deepsleep.

For a time during boot-up it must be OK for PA0 to be high, as long
as that high disappears before the program starts to run.  

What I am using is a DMOS driving a 1uF cap connected to PA0.  This
"works", but I suspect that maybe a tri-state buffer or ?? should be
used to drive PA0 correctly,

The one problem I have seen is  that the STM32F411 can be provoked into an "unwanted" state.
What can happen is that it goes to "who knows where" with a resulting
return to REPL, ie it stops running.

Any suggestions on how to keep the STM32F411 running and responding correctly
to a trigger signal would be appreciated. 

Edit:
I have been able to catch the waveform of the trigger pulse when I get this problem.
What is the pulse length suppose to be?  Using a 10X probe so that is 1V/div.  The rise time on a good trigger
is the same as a "bad" trigger.  Is the problem the 2nd pulse?  I would be happier missing a pulse rather than 
having the STM32F411 "bail-out" to REPL.

![20240909_090708](https://github.com/user-attachments/assets/012c3d9d-00cb-41cd-ac72-cf750dd9014e)

**Comment by bidrohini** [comment]
You can use an external watchdog timer to reset the system if it goes into an undesired state.

**Reply by davefes** [reply]
How do I detect the unknown state?  Perhaps to clarify the STM32F411 is in `deepsleep()` or `pyb.standby()`.  I have used a hardware WDT for other projects.

I struggled with similar issues on the ESP32.  As far as I can tell during deepsleep both the `machine.WDT()` and any timer used for a software watchdog stop running.  The only part-way solution I have used was to wake-up say every 1 minute "pat" the watchdogs and go back to sleep.  I am sure that I had issues with the ESP32s not coming out of deepsleep so my solution did not cover all possible fault conditions.

I realised that I don't really need deepsleep on STM32F411 as there is a sensor running all the time, which uses 4-5mA.  So, I tried using `pyb.stop()` ... now I need to work out how to get the unit to come out of "lightsleep", as PA0 only seems to work for triggering it out of deepsleep.

Edit:
Looking at this:
https://github.com/peterhinch/micropython-micropower/blob/master/UPOWER.md#5-module-ds_test
Maybe PC13 on the BlackPill
